pkg/kube/secretcontroller/secretcontroller.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secretcontroller
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/workqueue"
+
+	//"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	mcLabel    = "istio/multiCluster"
+	maxRetries = 5
+)
+
+// Required for unit test override.
+var loadKubeConfig = clientcmd.Load
+
+// addSecretCallback prototype for the add secret callback function.
+type addSecretCallback func(clientset kubernetes.Interface, dataKey string) error
+
+// removeSecretCallback prototype for the remove secret callback function.
+type removeSecretCallback func(dataKey string) error
+
+// Controller is the controller implementation for Secret resources
+type Controller struct {
+	kubeclientset  kubernetes.Interface
+	namespace      string
+	cs             *ClusterStore
+	queue          workqueue.RateLimitingInterface
+	informer       cache.SharedIndexInformer
+	addCallback    addSecretCallback
+	removeCallback removeSecretCallback
+}
+
+// RemoteCluster defines cluster structZZ
+type RemoteCluster struct {
+	secretName string
+}
+
+// ClusterStore is a collection of clusters
+type ClusterStore struct {
+	remoteClusters map[string]*RemoteCluster
+	storeLock      sync.RWMutex
```

Comment:
I don't think storeLock is needed. If I'm not wrong, then there are only two go routines here: the controller and the informer, and only the controller go routine changes this data structure.

Reply:
Agreed. Not sure why this code was needed in original implementation. Perhaps it's (pilot version) changed over time and is also not required. I'll remove.